crates/webidl/src/util.rs

```diff
@@ -173,192 +58,344 @@ fn result_ty(t: syn::Type) -> syn::Type {
     ty.into()
 }
 
-pub fn create_function<'a, I>(
-    name: &str,
-    arguments: I,
-    mut ret: Option<syn::Type>,
-    kind: backend::ast::ImportFunctionKind,
-    structural: bool,
-    catch: bool,
-) -> Option<backend::ast::ImportFunction>
-where
-    I: Iterator<Item = (&'a str, &'a webidl::ast::Type, bool)>,
-{
-    let rust_name = rust_ident(&name.to_snake_case());
-    let name = raw_ident(name);
-
-    let arguments = webidl_arguments_to_syn_arg_captured(arguments, &kind)?;
-
-    let js_ret = ret.clone();
-
-    if catch {
-        ret = Some(ret.map_or_else(|| result_ty(unit_ty()), |ret| result_ty(ret)))
-    }
-
-    let shim = {
-        let ns = match kind {
-            backend::ast::ImportFunctionKind::Normal => "",
-            backend::ast::ImportFunctionKind::Method { ref class, .. } => class,
-        };
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum TypePosition {
+    Argument,
+    Return,
+}
 
-        raw_ident(&format!("__widl_f_{}_{}", rust_name, ns))
-    };
+#[derive(Default)]
+pub struct FirstPass<'a> {
```

Comment:
I think a doc comment here explaining what the first pass over the WebIDL does, what information is collected, and for what use and why would be very valuable for future code readers and contributors.